pkg/machine/os/ostree.go

```diff
@@ -0,0 +1,132 @@
+type OSTree struct { //nolint:revive
+}
+
+// Apply takes an OCI image and does an rpm-ostree rebase on the image
+// If no containers-transport is specified,
+// apply will first check if the image exists locally, then default to pulling.
+// Exec-ing out to rpm-ostree rebase requires sudo, so this means apply cannot
+// be called within podman's user namespace if run as rootless.
+// This means that we need to export images in containers-storage to oci-dirs
+// We also need to do this via an exec, since if we tried to use the ABI functions,
+// we would enter the user namespace, the rebase command would fail.
+func (dist *OSTree) Apply(image string, opts ApplyOptions) error {
+	imageWithTransport := image
+
+	transport := alltransports.TransportFromImageName(image)
+
+	switch {
+	// no transport was specified
+	case transport == nil:
+		exists, err := execPodmanImageExists(image)
+		if err != nil {
+			return err
+		}
+
+		if exists {
+			fmt.Println("Pulling from", "containers-storage"+":", imageWithTransport)
+			dir, err := os.MkdirTemp("", pathSafeString(imageWithTransport))
+			if err != nil {
+				return err
+			}
+			err = os.Chmod(dir, 0755)
+			if err != nil {
+				return err
+			}
+			defer os.RemoveAll(dir)
+			err = execPodmanSave(dir, image)
+			if err != nil {
+				return err
+			}
+
+			imageWithTransport = "oci:" + dir
+		} else {
+			// if image doesn't exist locally, assume that we want to pull and use docker transport
+			imageWithTransport = "docker://" + image
+		}
+	// containers-transport specified
+	case transport.Name() == "containers-storage":
+		fmt.Println("Pulling from", image)
+		dir, err := os.MkdirTemp("", pathSafeString(strings.TrimPrefix(image, "containers-storage"+":")))
+		if err != nil {
+			return err
+		}
+		err = os.Chmod(dir, 0755)
+		if err != nil {
+			return err
+		}
+		defer os.RemoveAll(dir)
+		err = execPodmanSave(dir, image)
+		if err != nil {
+			return err
+		}
+		imageWithTransport = "oci:" + dir
+	}
+
+	ostreeCli := []string{"rpm-ostree", "--bypass-driver", "rebase"}
+	ostreeCli = append(ostreeCli, fmt.Sprintf("ostree-unverified-image:%s", imageWithTransport))
+
+	cmd := exec.Command("sudo", ostreeCli...)
+
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err := cmd.Run()
+
+	return err
+}
+
+// pathSafeString creates a path-safe name for our tmpdirs
+func pathSafeString(str string) string {
+	alphanumOnly := regexp.MustCompile(`[^a-zA-Z0-9]+`)
+
+	return alphanumOnly.ReplaceAllString(str, "")
+}
+
+// execPodmanSave execs out to podman save
+func execPodmanSave(dir, image string) error {
+	saveArgs := []string{"image", "save", "--format", "oci-dir", "-o", dir, image}
+
+	saveCmd := exec.Command("podman", saveArgs...)
+	saveCmd.Stdout = os.Stdout
+	saveCmd.Stderr = os.Stderr
+	err := saveCmd.Run()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// execPodmanSave execs out to podman image exists
+func execPodmanImageExists(image string) (bool, error) {
```

Comment:
does this run on the host or guest?

Reply:
Everything in ostree.go runs on the guest